include/dnscpp/tlsa.h

```diff
@@ -0,0 +1,154 @@
+
+        // The certificate used has a valid certification path pointing back to
+        // the certificate mention in this record, but there is no need for it
+        // to pass the PKIX certification path validation to a trusted root-CA.
+        trust_anchor_assertion = 2,
+
+        // The services uses a self-signed certificate. It is not signed by
+        // anyone else, and is exactly this record.
+        domain_issued_certificate = 3
+    };
+
+    /**
+     *  When connecting to the service and a certificate is received, the
+     *  selector field specifies which parts of it should be checked.
+     */
+    enum class Selector : uint8_t
+    {
+        // Select the entire certificate for matching.
+        entire_certificate = 0,
+
+        // Select just the public key for certificate matching. Matching the
+        // public key is often sufficient, as this is likely to be unique.
+        only_public_key = 1
+    };
+
+    /**
+     *  Matching type
+     */
+    enum class MatchingType : uint8_t
+    {
+        // The entire information selected is present in the certificate
+        // association data.
+        info_is_in_certificate_association_data = 0,
+
+        // Do a SHA-256 hash of the selected data.
+        do_SHA256 = 1,
+
+        // Do a SHA-512 hash of the selected data.
+        do_SHA512 = 2
+    };
+
+
+    /**
+     *  The constructor
+     *  @param  response        the response from which the record was extracted
+     *  @param  record          the record holding the CNAME
+     *  @throws std::runtime_error
+     */
+    TLSA(const Response &response, const Record &record) :
+        Extractor(record, TYPE_TLSA, 3)
+    {
+        const unsigned char *p = record.data();
+        _usage = static_cast<CertificateUsage>(*p++);
+        _selector = static_cast<Selector>(*p++);
+        _matching = static_cast<MatchingType>(*p++);
+        _association = p;
+    }
+
+    /**
+     *  Destructor
+     */
+    virtual ~TLSA() = default;
+
+    /**
+     *  How to verify the certificate?
+     *  @return the certificate usage
+     */
+    CertificateUsage certificateUsage() const { return _usage; }
+
+    /**
+     *  When a certificate is received, which parts of it should be checked?
+     *  @return the selector
+     */
+    Selector selector() const { return _selector; }
+
+    /**
+     *  How do we declare it a match?
+     *  @return the matching type
+     */
+    MatchingType matchingType() const { return _matching; }
+
+    /**
+     *  The actual data to be matched given the settings of the other fields.
+     *  @return a text string of hexadecimal data
+     */
+    const unsigned char *certificateAssociationData() const { return _association; }
+
+    /**
+     *  The length of the certificate assocation data.
+     * @return length
+     */
+    size_t certificateAssociationDataSize() const { return _record.size() - 3; }
+
+private:
+    CertificateUsage _usage;
```

Comment:
Codestyle has private members at the top

Reply:
And commented types.

Reply:
I cannot bring them all the way to the top because of the otherwise missing enum declarations. See 224b56e